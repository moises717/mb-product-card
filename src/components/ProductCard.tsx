import React, { createContext, } from 'react';

import { useProduct } from '../hooks/useProduct'
import { ProductContextProps, Product, onChangeArgs, initialValues, ProductCartHandler } from '../interfaces/interfaces';
import styles from '../styles/styles.module.css'

export interface Props {
    product: Product;
    //children?: React.ReactElement | React.ReactElement[];
    children: (args: ProductCartHandler) => JSX.Element | JSX.Element[];
    className?: string;
    style?: React.CSSProperties;
    onChange?: (args: onChangeArgs) => void;
    value?: number;
    initialValues?: initialValues
}

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {

    const { counter, increaseBy, maxCount, reset, isMaxCountReached } = useProduct({ onChange, product, value, initialValues, });

    return (
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount
        }}>
            <div className={`${styles.productCard} ${className}`}
                style={style}
            >
                {children({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product,
                    increaseBy,
                    reset
                })}
            </div>
        </Provider>
    )
}
