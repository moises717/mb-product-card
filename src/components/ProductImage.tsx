import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";

import noImg from '../assets/no-image.jpg'
import styles from '../styles/styles.module.css'


export interface Props {
    img?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductImage = ({ img = "", className, style }: Props) => {
    const { product } = useContext(ProductContext);

    let imageToShow: string;

    if (img) imageToShow = img;
    else if (product.img) imageToShow = product.img;
    else imageToShow = noImg;

    return (
        <img className={`${styles.productImg} ${className}`} src={imageToShow} alt="" style={style} />
    )
}