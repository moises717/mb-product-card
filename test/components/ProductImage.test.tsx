import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components/';
import { product2 } from '../data/products'


describe('ProductTitle', () => {
    test('Debe mostrar el componente correctamente con el titulo personalizado', () => {

        const wrapper = renderer.create(<ProductImage img='https://hola.png' />)

        expect(wrapper.toJSON()).toMatchSnapshot();


    })

    test('Debe de mostrar el componente con el nombre del producto', () => {

        const wrapper = renderer.create(
            <ProductCard product={product2} >
                {
                    () => <ProductImage img='https://hola.png' />
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();

    })
})

