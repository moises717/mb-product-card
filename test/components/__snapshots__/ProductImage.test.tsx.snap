// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle Debe de mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt=""
    className="productImg undefined"
    src="https://hola.png"
  />
</div>
`;

exports[`ProductTitle Debe mostrar el componente correctamente con el titulo personalizado 1`] = `
<img
  alt=""
  className="productImg undefined"
  src="https://hola.png"
/>
`;
